File: frontend/src/mocks/scans.ts
// ============================================================
// Mock 데이터 - 헬스 스캔 리포트 (Health Scans)
// 실제 API 연동 시 services/scanService.ts로 교체
// ============================================================

import type { HealthScan, BCSScore } from '@/types'

/** 데모용 헬스 스캔 리포트 목록 */
export const MOCK_SCANS: HealthScan[] = [
  // 뭉치(pet-001) - 최근 6회 스캔
  {
    id: 'scan-001',
    pet_id: 'pet-001',
    scan_date: '2024-06-01T14:30:00Z',
    bcs_score: 4 as BCSScore,
    gait_score: 88.5,
    eye_clarity_score: 92.0,
    voice_emotion_score: 76.3,
    ai_comment:
      '뭉치의 전반적인 건강 상태가 양호합니다. BCS 점수 4점으로 이상적인 체형을 유지하고 있으며, 보행 패턴도 정상 범위 내에 있습니다. 안구 상태도 맑고 건강합니다. 현재 식이 관리를 유지하시고 꾸준한 산책을 권장드립니다.',
    attachment_urls: ['s3://pet-id-uploads/scans/pet-001/scan-001.mp4'],
    scan_types: ['BCS', 'GAIT', 'AGING'],
    created_at: '2024-06-01T14:35:00Z',
  },
  {
    id: 'scan-002',
    pet_id: 'pet-001',
    scan_date: '2024-05-01T10:00:00Z',
    bcs_score: 5 as BCSScore,
    gait_score: 85.2,
    eye_clarity_score: 90.5,
    ai_comment:
      '지난달 대비 체중이 약간 증가했습니다. BCS 5점은 정상 범위이지만 약간 과체중 경향이 보입니다. 간식을 줄이고 운동량을 늘려주세요.',
    scan_types: ['BCS', 'GAIT'],
    created_at: '2024-05-01T10:05:00Z',
  },
  {
    id: 'scan-003',
    pet_id: 'pet-001',
    scan_date: '2024-04-01T11:00:00Z',
    bcs_score: 4 as BCSScore,
    gait_score: 90.1,
    eye_clarity_score: 93.2,
    ai_comment: '체형 및 보행 상태 모두 우수합니다. 노화 지표도 나이 대비 젊은 편입니다.',
    scan_types: ['BCS', 'AGING'],
    created_at: '2024-04-01T11:05:00Z',
  },
  {
    id: 'scan-004',
    pet_id: 'pet-001',
    scan_date: '2024-03-01T09:00:00Z',
    bcs_score: 3 as BCSScore,
    gait_score: 87.8,
    eye_clarity_score: 91.0,
    ai_comment: '약간 저체중 경향입니다. 칼로리 섭취량을 조금 늘려보세요.',
    scan_types: ['BCS'],
    created_at: '2024-03-01T09:05:00Z',
  },
  {
    id: 'scan-005',
    pet_id: 'pet-001',
    scan_date: '2024-02-01T14:00:00Z',
    bcs_score: 4 as BCSScore,
    gait_score: 89.3,
    eye_clarity_score: 92.5,
    ai_comment: '이상적인 체형을 유지하고 있습니다.',
    scan_types: ['BCS', 'GAIT'],
    created_at: '2024-02-01T14:05:00Z',
  },
  {
    id: 'scan-006',
    pet_id: 'pet-001',
    scan_date: '2024-01-01T10:00:00Z',
    bcs_score: 4 as BCSScore,
    gait_score: 91.0,
    eye_clarity_score: 94.1,
    ai_comment: '첫 번째 스캔 결과입니다. 모든 지표가 정상 범위에 있습니다.',
    scan_types: ['BCS', 'GAIT', 'AGING'],
    created_at: '2024-01-21T10:05:00Z',
  },

  // 두부(pet-002) - 최근 3회 스캔
  {
    id: 'scan-007',
    pet_id: 'pet-002',
    scan_date: '2024-06-05T15:00:00Z',
    bcs_score: 6 as BCSScore,
    gait_score: 78.4,
    eye_clarity_score: 85.2,
    voice_emotion_score: 82.0,
    ai_comment:
      '두부의 체형이 약간 과체중(BCS 6) 범위에 있습니다. 하루 식사량을 10~15% 줄이고 고단백 사료로 전환을 고려하세요. 안구 투명도는 정상이나 경과 관찰이 필요합니다.',
    scan_types: ['BCS', 'AGING', 'VOICE'],
    created_at: '2024-06-05T15:10:00Z',
  },
  {
    id: 'scan-008',
    pet_id: 'pet-002',
    scan_date: '2024-04-10T11:00:00Z',
    bcs_score: 5 as BCSScore,
    gait_score: 80.1,
    eye_clarity_score: 87.0,
    ai_comment: '체형은 정상 범위이나 약간 증가 추세입니다.',
    scan_types: ['BCS'],
    created_at: '2024-04-10T11:05:00Z',
  },
  {
    id: 'scan-009',
    pet_id: 'pet-002',
    scan_date: '2024-02-15T09:00:00Z',
    bcs_score: 5 as BCSScore,
    gait_score: 82.5,
    eye_clarity_score: 89.3,
    ai_comment: '전반적으로 건강한 상태입니다.',
    scan_types: ['BCS', 'GAIT'],
    created_at: '2024-02-15T09:05:00Z',
  },
]

/**
 * BCS 점수에 대한 설명 반환
 */
export function getBCSDescription(score: BCSScore): { label: string; color: string; description: string } {
  const descriptions: Record<BCSScore, { label: string; color: string; description: string }> = {
    1: { label: '극도 저체중', color: 'text-error-600', description: '갈비뼈, 척추, 골반이 명확히 돌출됨' },
    2: { label: '매우 저체중', color: 'text-error-500', description: '갈비뼈가 쉽게 만져짐, 근육 손실 있음' },
    3: { label: '저체중', color: 'text-warning-600', description: '갈비뼈가 적은 압력으로 만져짐' },
    4: { label: '이상적', color: 'text-secondary-600', description: '이상적인 체형 유지 중' },
    5: { label: '이상적 (약간 과체중)', color: 'text-secondary-500', description: '갈비뼈는 만져지나 지방이 약간 있음' },
    6: { label: '과체중', color: 'text-warning-500', description: '갈비뼈 위 지방이 과다함' },
    7: { label: '비만', color: 'text-warning-600', description: '갈비뼈가 만져지지 않음, 명확한 비만' },
    8: { label: '심한 비만', color: 'text-error-500', description: '전신 지방 과다, 운동 제한됨' },
    9: { label: '극심한 비만', color: 'text-error-600', description: '대규모 지방 축적, 즉시 식이 조절 필요' },
  }
  return descriptions[score]
}

/**
 * 반려동물의 스캔 목록 조회
 * 실제 API 연동 시 GET /pets/{pet_id}/scans 으로 교체
 */
export function mockGetScansByPetId(petId: string): HealthScan[] {
  return MOCK_SCANS
    .filter(s => s.pet_id === petId)
    .sort((a, b) => new Date(b.scan_date).getTime() - new Date(a.scan_date).getTime())
}

/**
 * 스캔 ID로 상세 조회
 * 실제 API 연동 시 GET /scans/{scan_id} 으로 교체
 */
export function mockGetScanById(scanId: string): HealthScan | null {
  return MOCK_SCANS.find(s => s.id === scanId) ?? null
}

/**
 * 신규 스캔 결과 Mock 생성 (AI 분석 시뮬레이션)
 * 실제 API 연동 시 POST /pets/{pet_id}/scans 으로 교체
 */
export function mockCreateScan(petId: string): HealthScan {
  const bcsScores: BCSScore[] = [3, 4, 4, 5, 5, 6]
  const randomBCS = bcsScores[Math.floor(Math.random() * bcsScores.length)]

  const newScan: HealthScan = {
    id: `scan-${Date.now()}`,
    pet_id: petId,
    scan_date: new Date().toISOString(),
    bcs_score: randomBCS,
    gait_score: Math.round((Math.random() * 30 + 70) * 10) / 10,
    eye_clarity_score: Math.round((Math.random() * 20 + 80) * 10) / 10,
    voice_emotion_score: Math.round((Math.random() * 30 + 60) * 10) / 10,
    ai_comment: `AI 분석이 완료되었습니다. BCS ${randomBCS}점으로 ${getBCSDescription(randomBCS).label} 상태입니다. ${getBCSDescription(randomBCS).description}`,
    scan_types: ['BCS', 'GAIT'],
    created_at: new Date().toISOString(),
  }
  MOCK_SCANS.unshift(newScan)
  return newScan
}

File: backend/chalicelib/controllers/scans.py
"""
Health Scans 컨트롤러
GET  /pets/{pet_id}/scans       — 시계열 스캔 목록 (페이지네이션)
POST /pets/{pet_id}/scans       — AI 분석 요청 → SQS push → 202 + job_id
GET  /scans/{scan_id}           — 특정 스캔 상세 리포트
"""
import json

from chalice import Blueprint

from chalicelib.auth.decorators import require_auth, require_pet_ownership
from chalicelib.core.db import get_session
from chalicelib.core.exceptions import AppError, NotFoundError
from chalicelib.core.helpers import fetch_many, fetch_one, insert
from chalicelib.core.response import accepted, ok, error as err_response
from chalicelib.models.scan import HealthScan
from chalicelib.services.sqs_service import push_scan_job

scans_bp = Blueprint(__name__)

_VALID_SCAN_TYPES = {"bcs", "gait", "eye_clarity", "voice_emotion"}


def _serialize(scan: HealthScan) -> dict:
    return {
        "id": scan.id,
        "pet_id": scan.pet_id,
        "scan_date": scan.scan_date,
        "job_status": scan.job_status,
        "bcs_score": scan.bcs_score,
        "gait_score": scan.gait_score,
        "eye_clarity_score": scan.eye_clarity_score,
        "voice_emotion_score": scan.voice_emotion_score,
        "ai_comment": scan.ai_comment,
        "scan_types": json.loads(scan.scan_types) if scan.scan_types else [],
        "attachment_urls": json.loads(scan.attachment_urls) if scan.attachment_urls else [],
        "created_at": str(scan.created_at),
    }


@scans_bp.route("/pets/{pet_id}/scans", methods=["GET"], cors=True)
@require_auth
@require_pet_ownership
def list_scans(pet_id):
    """
    Query: ?page=1&limit=20&sort=-scan_date
    Response: 시계열 스캔 목록
    """
    params = scans_bp.current_app.current_request.query_params or {}
    page = max(1, int(params.get("page", 1)))
    limit = min(100, max(1, int(params.get("limit", 20))))
    sort = params.get("sort", "-scan_date")

    with get_session() as session:
        result = fetch_many(session, HealthScan, page=page, limit=limit, sort=sort, pet_id=pet_id)
        return ok(
            [_serialize(s) for s in result["data"]],
            meta=result["meta"],
        )


@scans_bp.route("/pets/{pet_id}/scans", methods=["POST"], cors=True)
@require_auth
@require_pet_ownership
def create_scan(pet_id):
    """
    Body: { "s3_key": "health_scan/...", "scan_types": ["bcs", "gait"], "scan_date": "2026-04-08T10:00:00Z" }
    Response 202: { "job_id": "<scan_id>", "status": "PENDING" }
    """
    body = scans_bp.current_app.current_request.json_body or {}
    s3_key = (body.get("s3_key") or "").strip()
    scan_types = body.get("scan_types") or []
    scan_date = (body.get("scan_date") or "").strip()

    if not s3_key:
        return err_response("VALIDATION_ERROR", "s3_key는 필수입니다.", status_code=422)
    if not scan_types:
        return err_response("VALIDATION_ERROR", "scan_types는 1개 이상 필요합니다.", status_code=422)
    if not scan_date:
        from datetime import datetime, timezone
        scan_date = datetime.now(timezone.utc).isoformat()

    invalid = set(scan_types) - _VALID_SCAN_TYPES
    if invalid:
        return err_response(
            "VALIDATION_ERROR",
            f"유효하지 않은 scan_type: {', '.join(invalid)}. 허용값: {', '.join(_VALID_SCAN_TYPES)}",
            status_code=422,
        )

    try:
        with get_session() as session:
            scan = insert(
                session,
                HealthScan(
                    pet_id=pet_id,
                    scan_date=scan_date,
                    scan_types=json.dumps(scan_types),
                    attachment_urls=json.dumps([s3_key]),
                    job_status="PENDING",
                ),
            )
            job_id = scan.id

        push_scan_job(pet_id=pet_id, s3_key=s3_key, scan_types=scan_types, job_id=job_id)

        return accepted({"job_id": job_id, "status": "PENDING"})

    except AppError as e:
        return err_response(e.error_code, e.message, status_code=400)


def _get_scan_detail(scan_id: str, user: dict):
    """스캔 상세 조회 공통 로직. 두 경로에서 공유."""
    from chalicelib.models.pet import Pet

    try:
        with get_session() as session:
            scan = fetch_one(session, HealthScan, id=scan_id)

            # B2C 소유권 확인 (ADMIN 제외)
            if user["role"] not in {"ADMIN"}:
                pet = fetch_one(session, Pet, id=scan.pet_id)
                if pet.owner_id != user["user_id"]:
                    return err_response("AUTH_002", "해당 스캔에 대한 접근 권한이 없습니다.", status_code=403)

            return ok(_serialize(scan))
    except NotFoundError as e:
        return err_response(e.error_code, e.message, status_code=404)


@scans_bp.route("/scans/{scan_id}", methods=["GET"], cors=True)
@require_auth
def get_scan(scan_id):
    """API Spec §3.4 — 특정 스캔 상세 리포트."""
    user = scans_bp.current_app.current_request.context["user"]
    return _get_scan_detail(scan_id, user)


@scans_bp.route("/scan/report/{scan_id}", methods=["GET"], cors=True)
@require_auth
def get_scan_report(scan_id):
    """Backend-dev-instruction §6 경로 alias — /scans/{scan_id} 와 동일."""
    user = scans_bp.current_app.current_request.context["user"]
    return _get_scan_detail(scan_id, user)
